dev/com.ibm.ws.grpc.server/src/com/ibm/ws/grpc/GrpcApplicationManager.java

```diff
@@ -0,0 +1,184 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+
+package com.ibm.ws.grpc;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.container.service.annocache.AnnotationsBetaHelper;
+import com.ibm.ws.container.service.annotations.WebAnnotations;
+import com.ibm.ws.container.service.app.deploy.ApplicationInfo;
+import com.ibm.ws.container.service.app.deploy.WebModuleInfo;
+import com.ibm.ws.container.service.state.ApplicationStateListener;
+import com.ibm.ws.container.service.state.StateChangeException;
+import com.ibm.wsspi.adaptable.module.NonPersistentCache;
+import com.ibm.wsspi.adaptable.module.UnableToAdaptException;
+import com.ibm.wsspi.anno.targets.AnnotationTargets_Targets;
+
+import io.grpc.BindableService;
+import io.grpc.Server;
+
+@Component(service = {
+		ApplicationStateListener.class }, configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true)
+public class GrpcApplicationManager implements ApplicationStateListener {
+
+	private static final String CLASS_NAME = GrpcApplicationManager.class.getName();
+	private static final Logger logger = Logger.getLogger(GrpcApplicationManager.class.getName());
+
+	@Activate
+	protected void activate(ComponentContext cc) {
+	}
+
+	@Deactivate
+	protected void deactivate(ComponentContext cc) {
+	}
+
+	@Modified
+	protected void modified(Map<?, ?> newProperties) {
+	}
+
+	/**
+	 * Discover and start gRPC services provided by this application
+	 */
+	@Override
+	public void applicationStarting(ApplicationInfo appInfo) throws StateChangeException {
+		Set<String> grpcServiceClassNames = findGrpcServiceImplementors(appInfo);
+		Set<BindableService> services = initGrpcServices(appInfo, grpcServiceClassNames);
+		if (services != null && !services.isEmpty()) {
+			LibertyServerBuilder serverBuilder = new LibertyServerBuilder();
+			for (BindableService service : services) {
+				serverBuilder.addService(service.bindService());
+			}
+			Server server = serverBuilder.build();
+			try {
+				server.start();
+				ActiveGrpcServers.addServer(appInfo.getName(), server);
+			} catch (IOException e) {
+				logMessage(Level.FINE, "initGrpcServices",
```

Comment:
Do we want to do more than log the message?

Reply:
I've added a FFDC for now.  We'll go back over messages more carefully in another PR.